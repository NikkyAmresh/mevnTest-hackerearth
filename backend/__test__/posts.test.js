process.env.NODE_ENV = "test";
const db = require("../app/models/index");
const request = require("supertest");
const app = require("../app");
const jp = require("jsonpath");
const pointer = require("json-pointer");
const _ = require("lodash");

function getPointerPath(jsonPath) {
  return jsonPath.join("/").replace("$", "");
}

function matchSnapshot(val, ignorePaths = []) {
  const obj = _.cloneDeep(val);
  ignorePaths.forEach((ignorePath) => {
    jp.paths(obj, ignorePath).forEach((jsonPath) => {
      pointer.remove(obj, getPointerPath(jsonPath));
    });
  });

  expect(obj).toMatchSnapshot();
}

beforeAll((done) => {
  done();
});

afterAll((done) => {
  db.mongoose.connection.close();
  console.log("Disconnected from Mongoose");
  done();
});

let postID;

describe("POST /api/posts", () => {
  test("It should create a new post", async () => {
    const response = await request(app).post("/api/posts").send({
      title: "test title",
      description: "test description",
      published: false,
    });
    expect(response.statusCode).toBe(200);
    matchSnapshot(response.body, ["$.createdAt", "$.updatedAt", "id"]);
  });
});

describe("POST /api/posts", () => {
  test("It should create another new post", async () => {
    const response = await request(app).post("/api/posts").send({
      title: "test another title",
      description: "test another description",
      published: false,
    });
    expect(response.statusCode).toBe(200);
    matchSnapshot(response.body, ["$.createdAt", "$.updatedAt", "id"]);
  });
});

describe("GET /api/posts", () => {
  test("It should respond with an array of posts", async () => {
    const response = await request(app).get("/api/posts");
    expect(response.statusCode).toBe(200);
    matchSnapshot(response.body, [
      "$[*].createdAt",
      "$[*].updatedAt",
      "$[*].id",
    ]);
    if (response.body.length > 0) postID = response.body[0].id;
  });
});

describe("GET /api/posts/:id", () => {
  test("It should return new post", async () => {
    const response = await request(app).get(`/api/posts/${postID}`);
    expect(response.statusCode).toBe(200);
    matchSnapshot(response.body, ["$.createdAt", "$.updatedAt", "id"]);
  });
});

describe("PUT /api/posts/:id", () => {
  test("It should update the post", async () => {
    const response = await request(app).put(`/api/posts/${postID}`).send({
      title: "new Title",
      description: "new Description",
      published: true,
    });
    expect(response.statusCode).toBe(200);
    matchSnapshot(response.body);
  });
});

describe("DELETE /api/posts/:id", () => {
  test("It should delete the post", async () => {
    const response = await request(app).delete(`/api/posts/${postID}`);
    expect(response.statusCode).toBe(200);
    matchSnapshot(response.body);
  });
});

describe("DELETE /api/posts/:id", () => {
  test("It should delete all posts", async () => {
    const response = await request(app).delete(`/api/posts`);
    expect(response.statusCode).toBe(200);
    matchSnapshot(response.body);
  });
});

describe("Test a 404", () => {
  test("It should respond with a 404 status", async () => {
    const response = await request(app).get("/nowhere");
    expect(response.statusCode).toBe(404);
    matchSnapshot(response.text);
  });
});
