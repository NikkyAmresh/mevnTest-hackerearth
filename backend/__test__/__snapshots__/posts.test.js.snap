// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE /api/posts/:id It should delete all posts 1`] = `
Object {
  "message": "1 Posts were deleted successfully!",
}
`;

exports[`DELETE /api/posts/:id It should delete the post 1`] = `
Object {
  "message": "Post was deleted successfully!",
}
`;

exports[`GET /api/posts It should respond with an array of posts 1`] = `
Array [
  Object {
    "description": "test description",
    "published": false,
    "title": "test title",
  },
  Object {
    "description": "test another description",
    "published": false,
    "title": "test another title",
  },
]
`;

exports[`GET /api/posts/:id It should return new post 1`] = `
Object {
  "description": "test description",
  "published": false,
  "title": "test title",
}
`;

exports[`POST /api/posts It should create a new post 1`] = `
Object {
  "description": "test description",
  "published": false,
  "title": "test title",
}
`;

exports[`POST /api/posts It should create another new post 1`] = `
Object {
  "description": "test another description",
  "published": false,
  "title": "test another title",
}
`;

exports[`PUT /api/posts/:id It should update the post 1`] = `
Object {
  "message": "Post was updated successfully.",
}
`;

exports[`Test a 404 It should respond with a 404 status 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /nowhere</pre>
</body>
</html>
"
`;
